### PR TITLE
SE-153-Merge-external-into-internal

### PR DESCRIPTION
**Pull request by ShaneDuncan602**
You will probably have to filter out the md files.

### CODE REVIEW

File: src/data/PauseRules.sol [hunk 1]
```diff
@@ -91,7 +92,7 @@ contract PauseRules is IPauseRules, DataModule {
      * @dev Get the pauseRules data for a given tokenName.
      * @return pauseRules all the pause rules for the token
      */
-    function getPauseRules() external view onlyOwner returns (PauseRule[] memory) {
+    function getPauseRules() external view virtual onlyOwner returns (PauseRule[] memory) {
```
Comment (TJ-Everett) [inline src/data/PauseRules.sol:95]: This is missing the isPauseRulesEmpty() from the pause rules bool branch that was just merged with JB.

File: src/token/ProtocolERC20Handler.sol [hunk 2]
```diff
@@ -151,8 +147,15 @@ contract ProtocolERC20Handler is Ownable, ITokenHandlerEvents, AppAdministratorO
...
+                volumeTotalForPeriod,
+                totalSupplyForPeriod,
+                IToken(msg.sender).totalSupply(),
+                _to == address(0x00) ? int(_amount) * -1 : int(_amount),
+                lastSupplyUpdateTime
+            );
+            lastSupplyUpdateTime = uint64(block.timestamp);
         }
         //added the following lines to remove warnings TODO remove later
         _balanceFrom;
```
Comment (TJ-Everett) [inline src/token/ProtocolERC20Handler.sol]: Looks like the remove unused variables didn't get picked up either. I see in the history that this was added into the branch this morning but the variables are still here.

File: test/ApplicationAMM.t.sol [hunk 3]
```diff
@@ -838,9 +794,8 @@ contract ApplicationAMMTest is DiamondTestUtil, RuleProcessorDiamondTestUtil {
         // applicationAMM.swap(address(applicationCoin), 100);
 
         ///create new pause rule to check that swaps and trasnfers are paused
```
Comment (TJ-Everett) [inline test/ApplicationAMM.t.sol]: transfers**

File: test/ERC20AutoMintStaking.t.sol [hunk 4]
```diff
@@ -1,213 +0,0 @@
-// SPDX-License-Identifier: UNLICENSED
```
Comment (TJ-Everett) [inline test/ERC20AutoMintStaking.t.sol:1]: Are we removing all staking from internal intentionally?

**Review by TJ-Everett — CHANGES_REQUESTED** [review]
This looks great Shane. I notated that we are missing the pause rules bool and unused variables changes it seems. It also looks like we have completely removed staking from this branch, is that intentional?

File: test/ERC20AutoMintStaking.t.sol (hunk 4 above)
Comment (ShaneDuncan602) [reply test/ERC20AutoMintStaking.t.sol:1]: No, good find. It looks like this was deleted during the merge. I'll add them back.

File: .env.production [hunk 5]
```diff
@@ -26,10 +26,7 @@ ERC721_PRICING_CONTRACT=0x059CF723EFC126f82DC1D1bFC06EFa9bb3D1fa62
 ERC20_PRICING_CONTRACT=0x75afe612b0b6963f407aaBFbBc1096c1587456d9
 APPLICATION_AMM_HANDLER=0x3Aa96821A0c35DB49D1AAbe38413d058b9C862CB
 APPLICATION_ERC721_HANDLER=0x3F434D50520C642A53d703cb07F2c92121f6b549
-ERC20STAKING_CONTRACT=0x385eE9c70e460C9bFcBe08143D20aFad53D4139b
-ERC721STAKING_CONTRACT=0x8685A763F97b6228e4CF65F8B6993BFecc932e2b
-ERC20AUTOMINT_STAKING_CONTRACT=0xe1A8fd2b78f0F66698D057b8F55f9Cb862666401
-ERC721AUTOMINT_STAKING_CONTRACT=0x47456f745Ee61DA2181ba7336d96d78AC0707d5a
+
```
Comment (oscarsernarosero) [inline .env.production]: I think I saw Tayler asking about this, but yeah. I think GitHub assumes we are trying to get rid of the staking pieces which we are not trying to do. Idk how to solve that tho. Maybe trying to merge internal into external first in a separate branch, and then merge that branch back into internal? JB for sure knows a better way.

File: src/data/PauseRules.sol (hunk 1 above)
Comment (ShaneDuncan602) [reply src/data/PauseRules.sol:95]: Added the pause rule bool stuff

File: test/ApplicationAMM.t.sol (hunk 3 above)
Comment (ShaneDuncan602) [reply test/ApplicationAMM.t.sol]: done

File: test/ERC20AutoMintStaking.t.sol (hunk 4 above)
Comment (ShaneDuncan602) [reply test/ERC20AutoMintStaking.t.sol:1]: added them

**Review by oscarsernarosero — APPROVED** [review]
Looks good Shane

File: src/token/ProtocolERC20Handler.sol [hunk 6]
```diff
@@ -61,35 +54,31 @@ contract ProtocolERC20Handler is Ownable, ITokenHandlerEvents, AppAdministratorO
...
-    constructor(address _ruleProcessorProxyAddress, address _appManagerAddress, bool _upgradeMode) {
+    constructor(address _ruleProcessorProxyAddress, address _appManagerAddress, address _assetAddress, bool _upgradeMode) {
+        if (_appManagerAddress == address(0) || _ruleProcessorProxyAddress == address(0) || _assetAddress == address(0)) revert ZeroAddress();
         appManagerAddress = _appManagerAddress;
         appManager = IAppManager(_appManagerAddress);
         ruleProcessor = IRuleProcessor(_ruleProcessorProxyAddress);
+        transferOwnership(_assetAddress);
+        // register the supported interface, IAdminWithdrawalRuleCapable, ERC165
+        // _registerInterface(type(IAdminWithdrawalRuleCapable).interfaceId);
+        // _registerInterface(type(IAdminWithdrawalRuleCapable).interfaceId);
```
Comment (TJ-Everett) [inline src/token/ProtocolERC20Handler.sol]: Looks like this duplicated in the newest merge

**Review by TJ-Everett — APPROVED** [review]
Just one line of duped code otherwise this looks great!

File: .env.production (hunk 5 above)
Comment (ShaneDuncan602) [reply .env.production]: Added

File: src/token/ProtocolERC20Handler.sol (hunk 6 above)
Comment (ShaneDuncan602) [reply src/token/ProtocolERC20Handler.sol]: Fixed

**Comment by ShaneDuncan602** [comment]
Just to be sure that everything actually go into this branch, I went ahead and used a comparison tool(BeyondCompare) to check. It was also missing the example staking contract deployment scripts. I added them. Everything is merged, functioning, and ready.

**Pull request merged**